### PR TITLE
Do not shutdown service control when failed message ingestion fails

### PR DESCRIPTION
**Pull request by SzymonPobiega**

### CODE REVIEW

File: src/ServiceControl/Operations/ErrorIngestion.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        public ErrorIngestion(ErrorIngestor errorIngestor, string errorQueue, RawEndpointFactory rawEndpointFactory, IDocumentStore documentStore, LoggingSettings loggingSettings, Func<string, Exception, Task> onCriticalError)
+        {
+            this.errorIngestor = errorIngestor;
+            this.errorQueue = errorQueue;
+            this.rawEndpointFactory = rawEndpointFactory;
+            this.onCriticalError = onCriticalError;
+            importFailuresHandler = new SatelliteImportFailuresHandler(documentStore, loggingSettings, onCriticalError);
+        }
+
+        public async Task EnsureStarted()
```
Comment (danielmarbach) [inline src/ServiceControl/Operations/ErrorIngestion.cs]: Why not flow the cancellation token from watchdog into this method so that the waitasync could be cancelled if needed and we avoid indefinite hanging?

File: src/ServiceControl/Operations/ErrorIngestion.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            }
+            finally
+            {
+                startStopSemaphore.Release();
+            }
+        }
+
+        Task OnCriticalErrorAction(ICriticalErrorContext ctx) => onCriticalError(ctx.Error, ctx.Exception);
+
+        public async Task EnsureStopped()
```
Comment (danielmarbach) [inline src/ServiceControl/Operations/ErrorIngestion.cs]: Why not flow the cancellation token from watchdog into this method so that the waitasync could be cancelled if needed and we avoid indefinite hanging?

File: src/ServiceControl/Operations/ErrorIngestionCustomCheck.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        {
+            this.criticalErrorHolder = criticalErrorHolder;
+        }
+
+        public override Task<CheckResult> PerformCheck()
+        {
+            var failure = criticalErrorHolder.GetLastFailure();
+            if (failure == null)
+            {
+                return Task.FromResult(CheckResult.Pass);
```
Comment (danielmarbach) [inline src/ServiceControl/Operations/ErrorIngestionCustomCheck.cs]: Can be cached

File: src/ServiceControl/Operations/ErrorIngestionCustomCheck.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+            }
+
+            return Task.FromResult(CheckResult.Failed(failure));
```
Comment (danielmarbach) [inline src/ServiceControl/Operations/ErrorIngestionCustomCheck.cs]: If failure is cached this task can be as well

File: src/ServiceControl/Operations/ErrorIngestionCustomCheck.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        }
+
+        public class State
+        {
+            volatile string lastFailure;
```
Comment (danielmarbach) [inline src/ServiceControl/Operations/ErrorIngestionCustomCheck.cs]: If you make this a Task<string> you can optimize this for the customer. This would be actually an excellent case for ValueTask :(

File: src/ServiceControl/Operations/Watchdog.cs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+            watchdog = Task.Run(async () =>
+            {
+                while (!shutdownTokenSource.IsCancellationRequested)
+                {
+                    try
+                    {
+                        await ensureStarted().ConfigureAwait(false);
+                        clearFailure();
+                    }
+                    catch (OperationCanceledException)
```
Comment (danielmarbach) [inline src/ServiceControl/Operations/Watchdog.cs]: In which case would this raise such an exception given that this API doesn't officially support cancellation?

File: src/ServiceControl/Properties/launchSettings.json [hunk 7]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (danielmarbach) [inline src/ServiceControl/Properties/launchSettings.json]: Why is this needed?

File: src/ServiceControl/Hosting/Commands/ImportFailedErrorsCommand.cs [hunk 8]
```diff
@@ -26,15 +26,15 @@ public override async Task Execute(HostArguments args)
             var tokenSource = new CancellationTokenSource();
 
             var loggingSettings = new LoggingSettings(settings.ServiceName, LogLevel.Info, LogLevel.Info);
-            var bootstrapper = new Bootstrapper(ctx => { tokenSource.Cancel(); }, settings, busConfiguration, loggingSettings);
```
[the remark this answers is not in the document]
Comment (SzymonPobiega) [reply src/ServiceControl/Hosting/Commands/ImportFailedErrorsCommand.cs]: That means that if we get a critical error raised here, the process won't just stop automatically. This should not be a problem as if a message is failed to be re-imported, it stays in the database and there is no negative consequences to attempting to re-import them again.

File: src/ServiceControl/Operations/ErrorIngestion.cs [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+                if (ingestionEndpoint != null)
+                {
+                    return; //Already started
+                }
+
+                var rawConfiguration = rawEndpointFactory.CreateRawEndpointConfiguration(
+                    errorQueue,
+                    (messageContext, dispatcher) => errorIngestor.Ingest(messageContext));
+
+                rawConfiguration.Settings.Set("onCriticalErrorAction", (Func<ICriticalErrorContext, Task>)OnCriticalErrorAction);
```
[the remark this answers is not in the document]
Comment (SzymonPobiega) [reply src/ServiceControl/Operations/ErrorIngestion.cs]: Nope. This is unfortunately the only was to provide on critical error callback to NSB.Raw as it does not have a public API for this :( I need to talk to the maintainers of that piece of c...ode

File: src/ServiceControl/Operations/ErrorIngestionCustomCheck.cs (hunk 4 above)
Comment (SzymonPobiega) [reply src/ServiceControl/Operations/ErrorIngestionCustomCheck.cs]: I think as we expect this state to be not that common I think we should be fine with only caching the success result (for readability of the code).

File: src/ServiceControl/Operations/ErrorIngestionCustomCheck.cs (hunk 5 above)
Comment (SzymonPobiega) [reply src/ServiceControl/Operations/ErrorIngestionCustomCheck.cs]: You mean to have a `ValueTask` instead of `string` here?

File: src/ServiceControl/Operations/Watchdog.cs (hunk 6 above)
Comment (SzymonPobiega) [reply src/ServiceControl/Operations/Watchdog.cs]: You're right. But now that I added the cancellation token to `ensureStarted` it makes sense to have a catch here, right?

File: src/ServiceControl/Properties/launchSettings.json (hunk 7 above)
Comment (SzymonPobiega) [reply src/ServiceControl/Properties/launchSettings.json]: Committed by mistake :(

File: src/ServiceControl/Operations/Watchdog.cs (hunk 6 above)
Comment (danielmarbach) [reply src/ServiceControl/Operations/Watchdog.cs]: Then yes

File: src/ServiceControl/Operations/ErrorIngestionCustomCheck.cs (hunk 5 above)
Comment (danielmarbach) [reply src/ServiceControl/Operations/ErrorIngestionCustomCheck.cs]: yeah but we don't have it available so I would basically directly cache `Task<string>` and use `Task.FromResult` here
